lib/storage.js

```diff
@@ -0,0 +1,160 @@
+const path          = require("path")
+const marked        = require("marked");
+const Config        = require("../lib/config.js");
+const {parseAct}    = require("../lib/build.js");
+const {
+  getId,
+  testPath,
+  read,
+  revert,
+  render,
+  warn
+}                   = require("../lib/util.js")
+const KLAB_OUT      = process.env.KLAB_OUT || "out";
+const config_json   = JSON.parse(read("./config.json"));
+const config        = Config(config_json);
+const config_path   = config.src.specification;
+const rule_paths    = config.src.rules;
+const raw_md_config = read(config_path)
+const title         = path.basename(config_path, ".md");
+const raw_rules     = rule_paths.map(p => read(p)).join("\n\n")
+const rules         = marked
+  .lexer(raw_rules)
+  .filter(block => block.type === "code")
+  .map(block => block.text)
+  .join("\n\n")
+
+// returns the storage def object
+// Array of "lines"
+
+const getBlock = block => {
+  // clean Terminals
+  let T = str => str
+         .slice(1, -1)
+         .replace(/\[|\]|\(|\)|\.|\#|\$|\-|\_/g, e => "\\" + e)
+  // memorize Nonterminals
+  let NT = "(.*)"
+  // decide if `str` is terminal
+  let isT = str => /^\"[^\"]*\"$/.test(str)
+  // test syntax
+  let regex = "^"
+    + block.find(str => /^syntax/.test(str))
+    .split(" ")
+    // remove k junk
+    .slice(3, -1)
+    // reduce all Terminals and Nonterminals to a regex
+    .reduce((a, e, i) => a
+      // whitespace possible between args
+      + (i > 0 ? "\\s*" : "")
+      // add terminal or memorize NT
+      + (isT(e) ? T(e) : NT)
+      , "")
+    + "$";
+  let test = str => (new RegExp(regex)).test(str)
+  let getArgs = str => str.match(new RegExp(regex)).slice(1)
+  let act = str => (block
+    .find(line => /^\s*\/\/\sact\:/.test(line)) || "")
+    .replace(/^\s*\/\/\sact\:\s*/, "")
+    .replace(/\$(\d+)/g, (_, d) => getArgs(str)[d])
+  let doc = str => (block
+    .find(line => /^\s*\/\/\sdoc\:/.test(line)) || "")
+    .replace(/^\s*\/\/\sdoc\:\s*/, "")
+    .replace(/\$(\d+)/g, (_, d) => getArgs(str)[d])
+  return {
+    test,
+    getArgs,
+    act,
+    doc
+  }
+}
+
+const storage_def = rules
+  .split(/\n(\ \t)*\n/)
+  .filter(block => block !== undefined)
+  .filter(block => /^(\s*\/\/[^\n]*\n)*syntax/.test(block))
+  .map(block => block.split("\n"))
+  .reduce((a, block) => {
+    return a.concat(getBlock(block));
+  }, [])
+const getStorageDef = str => storage_def.find(def => def.test(str))
+
+//TODO: refactor and get rid of these repetitions
+const mapInterface = (is, str) => {
+  is.forEach(([t, n]) => {
+    str = str.replace(new RegExp("([^\\w\.]|^)" + n + "([^\\w]|$)", "g"), (a, b, c, d) => {
+      return b + "ABI_" + n + c;
+    })
+  })
+  return str;
+}
+
+const toK = str => str
```

Comment:
this is redundant with lib/build.js
can we somehow unify them in either utils.js or kast.js or another one?!